### PR TITLE
"Live" mode for faster development

### PR DESCRIPTION
**Pull request by jimmc**
A few changes to make development a little easier.

### CODE REVIEW

File: sources/web/datalab/config/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Running with --live, the container creates a metadata file here.
+metadata.json
```
Comment (yebrahim) [inline sources/web/datalab/config/.gitignore]: I think this file is obsolete anyway, I can't see where we use it. This is separate from this change though.

File: containers/datalab/build.sh [hunk 2]
```diff
@@ -25,13 +25,21 @@ VERSION_SUBSTITUTION="s/_version_/0.5.$VERSION/"
 COMMIT=`git log --pretty=format:'%H' -n 1`
 COMMIT_SUBSTITUTION="s/_commit_/$COMMIT/"
 
+if [ -z "$1" ]; then
+  pythonPath=''
```
Comment (jimmc) [inline containers/datalab/build.sh]: Done. Yeah, I didn't choose a very good name for that.

File: containers/datalab/run.sh [hunk 3]
```diff
@@ -20,27 +20,68 @@
...
-    ENTRYPOINT="/bin/bash"
-  fi
-fi
+DEVROOT_DOCKER_OPTION=''
+while [ $# -gt 0 ]; do
+  case "$1" in
+    shell)
+      ENTRYPOINT="/bin/bash"
+      ;;
+    --live)
```
Comment (jimmc) [inline containers/datalab/run.sh]: Done.

File: containers/datalab/run.sh [hunk 4]
```diff
@@ -20,27 +20,68 @@
...
+      # development directory so your changes are visible to the container
+      # without rebuilding.
+      DEVROOT=$(cd "$HERE" && realpath ../..)
+      DEVROOT_DOCKER_OPTION="-v $DEVROOT:/devroot"
+      echo "Updating build-dev"
+      rsync -a \
+          --exclude web/nb/config \
+          --exclude web/nb/static \
+          --exclude web/nb/templates \
+          $DEVROOT/build/ $DEVROOT/build-dev
```
Comment (jimmc) [inline containers/datalab/run.sh]: Done.

File: containers/datalab/run.sh [hunk 5]
```diff
@@ -20,27 +20,68 @@
...
+      if [ ! -h $DEVROOT/build-dev/web/nb/node_modules ]; then
+        echo linking node_modules
+        ln -s /datalab/web/node_modules $DEVROOT/build-dev/web/nb/node_modules
```
Comment (jimmc) [inline containers/datalab/run.sh]: There are a whole bunch of symlinks in the build-dev directory. I don't think it's worth doing any partial clean up.

File: containers/datalab/run.sh [hunk 6]
```diff
@@ -20,27 +20,68 @@
...
+      # The files in the subdirs of build/web/nb are just copied from the
+      # source dirs. In order to allow using updated versions of those files
+      # without having to run the build script, we remove them from build-dev
+      # and link them directly to the source dirs.
+      # You need to restart the notebook server in the container to make it
+      # pick up your changes.
+      for dir in config static templates; do
+        if [ ! -h $DEVROOT/build-dev/web/nb/$dir ]; then
+          echo linking $dir
+          ln -s /devroot/sources/web/datalab/$dir $DEVROOT/build-dev/web/nb/$dir
```
Comment (jimmc) [inline containers/datalab/run.sh]: No, it should be /devroot. Although this symlink is being created on the workstation, it is intended to be resolved from within the container, where the /devroot directory exists. I added a comment to the code.

**Comment by jimmc** [comment]
Re using the build directory, that was the first thing I tried. I had trouble managing the symlinks, I was getting errors during the container build. I switched to build-dev so that the build directory could continue to be optimized for the build process, and the build-dev directory could be hacked up at will to make live-mode work.

**Comment by jimmc** [comment]
I think we can get auto-reload functionality by adding the "--watch" option to the forever command. When I tried that, the node server would auto-restart when I changed a file, and I could just do a page reload and get the new code. Let me know if you think I should add that.

**Comment by yebrahim** [comment]
Does running an `npm install` in the build directory inside the container resolve those references?

**Comment by jimmc** [comment]
This PR has been superseded by the simpler and better implementation in #1321.

**Pull request closed without merging**